quinn/src/incoming_connection.rs

```diff
@@ -0,0 +1,149 @@
+use std::{
+    fmt,
+    future::{Future, IntoFuture},
+    net::{IpAddr, SocketAddr},
+    pin::Pin,
+    task::{Context, Poll},
+};
+
+use bytes::BytesMut;
+use proto::ConnectionError;
+use thiserror::Error;
+
+use crate::{
+    connection::{Connecting, Connection},
+    endpoint::EndpointRef,
+};
+
+/// An incoming connection for which the server has not yet begun its part of the handshake
+pub struct IncomingConnection(Option<State>);
+
+struct State {
+    inner: proto::IncomingConnection,
+    endpoint: EndpointRef,
+    response_buffer: BytesMut,
+}
+
+impl IncomingConnection {
+    pub(crate) fn new(
+        inner: proto::IncomingConnection,
+        endpoint: EndpointRef,
+        response_buffer: BytesMut,
+    ) -> Self {
+        Self(Some(State {
+            inner,
+            endpoint,
+            response_buffer,
+        }))
+    }
+
+    /// The local IP address which was used when the peer established
+    /// the connection
+    pub fn local_ip(&self) -> Option<IpAddr> {
```

Comment:
Nit: these simple getters should live below the consuming functions (`accept()` et al).

Reply:
Done.